### ISSUE TITLE
BaseConnection _handle_error raises socket.timeout but nobody handles it

### ISSUE THREAD

**Issue by adrianheron** [pika/pika#674]
I ran into an issue where it looks like a .recv call raised a ETIMEDOUT exception, which _handle_error turned into a socket.timeout exception.  Here is a partial stack trace:
```
<class '_socket.timeout'>
Traceback (most recent call last):
...  
File ".../site-packages/pika/adapters/base_connection.py", line 364, in _handle_events
    self._handle_read()
  File ".../site-packages/pika/adapters/base_connection.py", line 407, in _handle_read
    return self._handle_error(error)
  File ".../site-packages/pika/adapters/base_connection.py", line 309, in _handle_error
    raise socket.timeout
timeout
```
If I understand things correctly, the ioloop as woken up the connection saying the socket was readable, _handle_read tries to read from the socket and generates a ETIMEDOUT exception, this gets caught and passed to _handle_error, _handle_error then raises a socket.timeout exception which isn't caught and causes a crash.

Presumably the original error was some random network problem causing the socket to disconnect which is fair enough, but the generated exception doesn't seem to be caught by anywhere that calls _handle_error().

So my question has two parts:
1. What is the intention with raising the socket.timeout exception, should it be caught somewhere?
2. Would the correct behaviour in the event of ETIMEDOUT be to call self._handle_timeout() or self._handle_disconnect()?

**Comment by vitaly-krugl** [comment]
@adrianheron: In pika master, the socket is configured to be non-blocking after connection is established, so you should never see a socket timeout error after that. During connection establishment, a different error-handling strategy is used, so it would never end up in `_handle_error` during connection establishment.

Please provide the following additional information:

1. What's the platform
2. Which version of pika are you using and on which python version?
3. Steps to reproduce (if known/reproducible)
4. Full traceback of the exception
5. pika log output  of the failing session

**Comment by vitaly-krugl** [comment]
@adrianheron, would you mind attaching a complete traceback?

**Comment by adrianheron** [comment]
I'm afraid that the full stack trace doesn't provide anything other than I've already posted:
```python
<class '_socket.timeout'>
Traceback (most recent call last):
  File "...", line 101, in <my_main_function>()
  File "...", line 221, in <my_io_loop>()
  File "/usr/lib/pypy/lib_pypy/_functools.py", line 22, in __call__
    return self.func(*(self.args + fargs), **fkeywords)
  File "...", line 124, in fire_fd_handler
    self._fd_handlers[fileno](fileno, event)
  File ".../pypy/site-packages/pika/adapters/base_connection.py", line 364, in _handle_events
    self._handle_read()
  File ".../pypy/site-packages/pika/adapters/base_connection.py", line 407, in _handle_read
    return self._handle_error(error)
  File ".../pypy/site-packages/pika/adapters/base_connection.py", line 309, in _handle_error
    raise socket.timeout
timeout
```

Perhaps the one strange thing is that it is running through pypy, but I'm not sure how that would relate to this specific error.

The pika version is 0.10.0.

I am unable to reproduce the error (it has happened three times over the last several weeks that I'm aware of) so I cannot generate debug-level pika logs.

One thing to note is that the original exception generated in _handle_read was not a socket.timeout exception, it is _handle_error that explicitly raises a socket.timeout exception in response to the original exception.  I'm not sure if that sheds any light on this.

**Comment by vitaly-krugl** [comment]
@adrianheron: Last night, I verified the issue as follows, using BlockingConnection (which is now implemented on top of the SelectConnection/BaseConnection/Connection stack):

1. Connect to Rabbit MQ on remote machine
2. Disconnect client from network (pull the Ethernet cable)
3. Attempt to create a channel on that connection.

This triggered `error(60, 'Operation timed out')` when the socket tried to send. On pika 0.10.0, this results in `socket.timeout` from `_handle_error` just as you indicated.

There is already a fix for this issue in master via PR #685.

https://github.com/pika/pika/pull/685/files#diff-1a5308d8b684fdafac1f89e376e8933dL326
